[Merrickneo] iP

Merrickneo opened the pull request:
# DukeProject
> Let _Mr Duke_ help you declutter your tasks to help you get on top of things.

## Types of tasks :
1. ToDos
2. Deadline
3. Events

## Features 
- 📑 List tasks
- ❌  Delete tasks
- Ⓜ️ Mark/Unmark tasks

Commit:
Add Gradle support

Merrickneo committed:
Update .gitignore to prevent .class files from being committed

Merrickneo committed:
Add greeting, echo and exit feature under Level-1 iP increment

Merrickneo committed:
Add Task class to store information about the details of each Task (Level 2)

Merrickneo committed:
Add task storing feature for Duke (Level 2)

Merrickneo committed:
Add feature to mark and unmark task's completion status

Merrickneo committed:
Add various definitions of various subclasses of Tasks (Deadline, Event, ToDo)

Merrickneo committed:
Add TaskTracker class to keep track of all Tasks

Merrickneo committed:
Update Duke main class to support the various initialisations of subclasses of Tasks

Merrickneo committed:
Adjust divider between texts

Merrickneo committed:
Add automated UI testing

Merrickneo committed:
Add DukeInputError Exception to store error message

Merrickneo committed:
Create static method in Duke class to handle command input and add exception handling.

Add command input handling method and exception handling to Duke.

Merrickneo committed:
Add delete task feature.

Merrickneo committed:
Generalise exception handling for "mark", "unmark" and "delete" operations.

Merrickneo committed:
Remove redundant methods in TaskTracker.

Merrickneo committed:
Update exception string message

Merrickneo committed:
Add save functionality whenever there is a change to the list.

Merrickneo committed:
Update gitignore file

Merrickneo committed:
Merge branch 'branch-Level-7'

* branch-Level-7:
  Add save functionality whenever there is a change to the list.

Merrickneo committed:
Update DeadlineTask and Event classes to handle dates and times.

Merrickneo committed:
Merge branch 'branch-Level-8'

* branch-Level-8:
  Update DeadlineTask and Event classes to handle dates and times.

Merrickneo committed:
Update name of TaskTracker to TaskList

Merrickneo committed:
Add more classes to encapsulate Ui, Storage, Command Parsing and TaskList portions of the program

Merrickneo committed:
Rename DukeInputError to DukeException

Merrickneo committed:
Update save and load task feature

Merrickneo committed:
Refactor files

Merrickneo committed:
Debug load feature for Tasks

Merrickneo committed:
Merge remote-tracking branch 'upstream/add-gradle-support'

* upstream/add-gradle-support:
  Add Gradle support

Merrickneo committed:
Add gradle support to check style

Merrickneo committed:
Add Unit Testing using Gradle for TaskList and DateTimeParser

Merrickneo committed:
Add JavaDoc for all classes.

Merrickneo committed:
Merge branch 'A-JavaDoc'

* A-JavaDoc:
  Add JavaDoc for all classes.

Merrickneo committed:
Edited code to adhere to coding standards.

Merrickneo committed:
Merge branch 'A-CodingStandard'

* A-CodingStandard:
  Edited code to adhere to coding standards.

Merrickneo committed:
Add find functionality to search for tasks that have the keyword specified.

Merrickneo committed:
Merge branch 'branch-Level-9'

* branch-Level-9:
  Add find functionality to search for tasks that have the keyword specified.

Merrickneo committed:
Update Task.java file.

Merrickneo committed:
Merge branch 'branch-Level-9'

* branch-Level-9:
  Update Task.java file.

PROGRAMMERHAO commented on src/main/java/duke/DeadlineTask.java:
```diff
@@ -0,0 +1,91 @@
+     * Outputs the deadline of the DeadlineTask in a String format
+     * @return String format of the DeadlineTask's deadline.
+     */
+    public String dateTimeFormatter() {
```
Perhaps this could be named as a verb as well? Something like formatDateTime.

PROGRAMMERHAO commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,88 @@
+     * @param keyword
+     * @return Boolean indicating if the keyword is found in the task description.
+     */
+    public boolean containsKeyword(String keyword) {
```
I like how you use boolean-like name here!

PROGRAMMERHAO commented on src/main/java/util/Storage.java:
```diff
@@ -0,0 +1,60 @@
+    public void saveTasks(TaskList tasks) throws DukeException {
+        try {
+            FileWriter fw = new FileWriter(this.filePath);
+            String tl = tasks.saveTaskList();
```
Maybe still could have used a more descriptive name here.

nicleejy commented on src/main/java/duke/DeadlineTask.java:
```diff
@@ -0,0 +1,91 @@
+     * @param t TaskList to store the newly created DeadlineTask
+     * @throws DukeException If command is invalid.
+     */
+    public static void createDeadlineTask(String command, TaskList t) throws DukeException {
+        ArrayList<String> input = new ArrayList<>(Arrays.asList(command.split(" ")));
+        if (input.size() <= 1) {
+            throw new DukeException("deadline");
+        }
+        int byIndex = input.indexOf("/by");
+        StringBuilder taskName = new StringBuilder();
+        StringBuilder deadline = new StringBuilder();
+        for (int i = 1; i < input.size(); i++) {
+            if (i < byIndex) {
+                taskName.append(input.get(i));
+                if (i < byIndex - 1) {
+                    taskName.append(" ");
+                }
+            } else if (i > byIndex) {
+                deadline.append(input.get(i));
+                if (i < input.size() - 1) {
+                    deadline.append(" ");
+                }
+            }
+        }
+        DeadlineTask d = new DeadlineTask(taskName.toString(), DateTimeParser.dateTimeParser(deadline.toString()));
+        t.addTask(d);
+    }
```
I think it's a great design decision to put the user input processing specific to a deadline within the `DeadlineTask` class itself. 👍

nicleejy commented on src/main/java/util/DukeException.java:
```diff
@@ -0,0 +1,42 @@
+     * Constructor of DukeException.
+     * @param inputType Description of error.
+     */
+    public DukeException(String inputType) {
+        if (taskInput.contains(inputType)) {
+            this.errorMessage = String.format("The description of a %s cannot be empty", inputType);
+        } else if (markInput.contains(inputType)) {
+            this.errorMessage = String.format("    Usage: '%s taskNumber'", inputType);
+        } else if (inputType.equals("bounds")) {
+            this.errorMessage = "I'm sorry, but please specify a valid task index.";
+        } else {
+            this.errorMessage = inputType;
+        }
+    }
```
I really like this method of checking for the invalid input and returning a corresponding error message. Will want to learn from this when creating exception classes in the future. 👍

nicleejy commented on src/main/java/util/Storage.java:
```diff
@@ -0,0 +1,60 @@
+     * @return BufferedReader object that contains the list of tasks.
+     * @throws DukeException If IOException from reading duke.txt is encountered.
+     */
+    public BufferedReader load() throws DukeException {
```
Might be better for the method to have a more descriptive name, such as `loadData()` or  `loadTasksFromFile`.

nicleejy commented on src/main/java/util/TaskList.java:
```diff
@@ -0,0 +1,167 @@
+                String taskType = taskArr[TASK_TYPE];
+                switch (taskType) {
+                case "T":
+                    boolean completion = Boolean.parseBoolean(taskArr[IS_DONE]);
```
Would be better to rename this boolean variable to `isCompleted` or `hasCompleted`

Merrickneo replied:
Thanks for the spot!

nicleejy commented on src/main/java/util/Ui.java:
```diff
@@ -0,0 +1,54 @@
+        System.out.println(divider);
+        System.out.println("    Hello! I'm Duke");
+        System.out.println("    What can I do for you?");
+        System.out.println(divider);
```
The Ui class is a great separation and makes for more organised code!

nicleejy reviewed (commented):
Overall the code quality is great and organised, and matches the Java coding standard 👍 Just made some comments on minor issues like name of variables and methods. Do let me know if you have any questions :)

PROGRAMMERHAO commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,98 @@
+                if (i < toIndex) {
+                    start += " ";
+                }
+            } else if (i > toIndex) {
```
Good to use else-if here to make the code looks cleaner!

PROGRAMMERHAO commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,98 @@
+    }
+
+    /**
+     * Static method to create a Event Task from user input.
```
Maybe could start by directly saying "creates...."

PROGRAMMERHAO reviewed (commented):
The overall project is great, and there are only some minor details related to naming or javadoc that could have been changed a bit. The code in each file is very clean and if-else statements are used to make the structure look organised. LGTM!

Merrickneo committed:
Remove extra spacing in Task.java

Merrickneo committed:
Add GUI functionality to Duke.

Merrickneo committed:
Merge branch 'branch-Level-10'

* branch-Level-10:
  Add GUI functionality to Duke.

Merrickneo committed:
Update build.gradle file to include JavaFX libraries

Build.gradle file does not include JavaFX libraries

./gradlew shadowJar does not bundle JavaFX third-party libraries into the jar file

Add JavaFX libraries dependencies to build.gradle

Merrickneo committed:
Add assertions to test assumptions of the state of Duke

Merrickneo committed:
Merge pull request #1 from Merrickneo/branch-A-Assertions

Add assertions to test assumptions of the state of Duke

Merrickneo committed:
Update code to follow Java Coding Standards

Merrickneo committed:
Merge pull request #2 from Merrickneo/branch-A-CodeQuality

Update code to follow Java Coding Standards

Merrickneo committed:
Modify output for list task feature

Merrickneo committed:
Add gradle.yml file to set up Continuous Integration (CI)

Merrickneo committed:
Improve code quality

Merrickneo committed:
Add snooze feature to iP

Merrickneo committed:
Merge pull request #3 from Merrickneo/B-Snooze

Add snooze feature to iP

Merrickneo committed:
Merge branch 'master' of https://github.com/Merrickneo/ip

* 'master' of https://github.com/Merrickneo/ip:
  Add snooze feature to iP

Merrickneo committed:
Update DeadlineTask.java

Merrickneo committed:
Merge branch 'master' into B-Snooze-Temp

Merrickneo committed:
Merge pull request #4 from Merrickneo/B-Snooze-Temp

Update DeadlineTask.java

Merrickneo committed:
Merge branch 'master' of https://github.com/Merrickneo/ip

* 'master' of https://github.com/Merrickneo/ip:
  Update DeadlineTask.java
  Merge branch 'master' of https://github.com/Merrickneo/ip

Merrickneo committed:
Update DeadlineTask

Merrickneo committed:
Debug snooze feature

Merrickneo committed:
Update file path for testing scipts

Merrickneo committed:
Update Javadocs

Merrickneo committed:
Edit Ui class logic

Merrickneo committed:
Add product screenshot of Duke

Merrickneo committed:
Format code to comply with coding standards

Merrickneo committed:
Remove extra spaces in output text

Merrickneo committed:
Adjust import order for testing scripts

Merrickneo committed:
Update assertions for JUnit Testing

Merrickneo committed:
Update README.md

Merrickneo committed:
Add header line for findTask() output

Merrickneo committed:
Remove hardcoded filepath for Duke

Merrickneo committed:
Update import order for Duke.java
